Remotion/Web/Development.Analyzers/RMWEB0001_WebStringAnalyzer.cs

```diff
@@ -0,0 +1,223 @@
+                ctx => AnalyzeInvocationExpression (ctx, symbolContext),
+                ImmutableArray.Create (SyntaxKind.InvocationExpression));
+          });
+    }
+
+    private void AnalyzeInvocationExpression (SyntaxNodeAnalysisContext ctx, SymbolContext symbols)
+    {
+      var invocationSymbolInfo = ctx.SemanticModel.GetSymbolInfo (ctx.Node);
+
+      if (invocationSymbolInfo.Symbol == null)
+        return;
+
+      var node = (InvocationExpressionSyntax) ctx.Node;
+      var invocationOriginalDefinitionSymbol = invocationSymbolInfo.Symbol.OriginalDefinition;
+
+      if (SymbolEqualityComparer.Default.Equals (invocationOriginalDefinitionSymbol, symbols.StringBuilderAppendMethodSymbol))
+      {
+        var argumentType = GetFirstArgumentTypeSymbol (node, ctx.SemanticModel);
+        if (SymbolEqualityComparer.Default.Equals (argumentType, symbols.WebStringTypeSymbol)
+            || SymbolEqualityComparer.Default.Equals (argumentType, symbols.PlainTextStringTypeSymbol))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              symbols.StringBuilderAppendMethodSymbol.ToString(),
+              argumentType.ToString(),
+              "Call '.ToString(WebStringEncoding.HtmlWithTransformedLineBreaks)' on the WebString instance instead.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+      else if (SymbolEqualityComparer.Default.Equals (invocationOriginalDefinitionSymbol, symbols.HtmlTextWriterWriteMethodSymbol))
+      {
+        var argumentType = GetFirstArgumentTypeSymbol (node, ctx.SemanticModel);
+        if (SymbolEqualityComparer.Default.Equals (argumentType, symbols.WebStringTypeSymbol)
+            || SymbolEqualityComparer.Default.Equals (argumentType, symbols.PlainTextStringTypeSymbol))
+        {
+          var diagnostic = Diagnostic.Create (
+              DiagnosticDescriptor,
+              ctx.Node.GetLocation(),
+              symbols.HtmlTextWriterWriteMethodSymbol.ToString(),
+              argumentType.ToString(),
+              $"Use '{argumentType.ToString()}.Write(HtmlTextWriter)' instead.");
+          ctx.ReportDiagnostic (diagnostic);
+        }
+      }
+      else if (AnyWithResult (
+          symbols.StringJoinMethodSymbols,
+          s => SymbolEqualityComparer.Default.Equals (invocationOriginalDefinitionSymbol, s),
+          out var methodSymbol))
+      {
+        var argumentTypes = GetArgumentTypeSymbols (node, ctx.SemanticModel);
+        if (AnyWithResult (
+            argumentTypes,
+            a => SymbolEqualityComparer.Default.Equals (a, symbols.WebStringTypeSymbol)
+                 || SymbolEqualityComparer.Default.Equals (a, symbols.PlainTextStringTypeSymbol)
+                 || ctx.Compilation.ClassifyConversion (a, symbols.WebStringEnumerable).IsImplicit
+                 || ctx.Compilation.ClassifyConversion (a, symbols.PlainTextStringEnumerable).IsImplicit,
```

Comment:
I'm learning that it's not so easy to check "correctness" during the review when you're not fluent in Roslyn.

I think it would helpful for maintainability to add a short summary to each of the IF blocks with a desciption of the cases covered by the block. Otherwise, you have to reverse engineer the cases from the tickets that belong to the commit and by jumping to (finding) and then reading the tests just to understand what's the intent.

Another option is always to factor out methods that encapsulate each use case.

Thoughts? Pro/Con?

Comment example:
// Detects string.Join (string, (object) WebString)
// Detects string.Join (string, (object) PlainTextString)

Reply:
I thought about adding comments but it seemed clear enough - but its not. I'll look into it.